Multithreaded calls to remote EJB may result in LinkageError

# Description #
----------

When remote EJB methods are called with multiple threads a LinkageError may occur.

## Expected Outcome

Remote EJB methods may be called without LinkageError.

## Current Outcome

```
Caused by: java.lang.LinkageError: loader &#40;instance of  org/glassfish/web/loader/WebappClassLoader&#41;: attempted  duplicate class definition for name: &quot;de/edict/eoc/server/proxy/_CasinoEnvironmentConstructorApi5_Remote&quot;
	at sun.misc.Unsafe.defineClass&#40;Native Method&#41;
	at org.glassfish.pfl.basic.reflection.BridgeBase.defineClass&#40;BridgeBase.java:227&#41;
	at org.glassfish.pfl.dynamic.codegen.impl.CodeGeneratorUtil.makeClass&#40;CodeGeneratorUtil.java:70&#41;
	at org.glassfish.pfl.dynamic.codegen.spi.Wrapper._generate&#40;Wrapper.java:1083&#41;
	at org.glassfish.pfl.dynamic.codegen.spi.Wrapper._generate&#40;Wrapper.java:1067&#41;
	at com.sun.ejb.EJBUtils.generateAndLoad&#40;EJBUtils.java:599&#41;
	at com.sun.ejb.EJBUtils.loadGeneratedRemoteBusinessClasses&#40;EJBUtils.java:513&#41;
	at com.sun.ejb.EJBUtils.loadGeneratedRemoteBusinessClasses&#40;EJBUtils.java:471&#41;
	at com.sun.ejb.containers.RemoteBusinessWrapperBase.readObject&#40;RemoteBusinessWrapperBase.java:122&#41;
	at sun.reflect.NativeMethodAccessorImpl.invoke0&#40;Native Method&#41;
	at sun.reflect.NativeMethodAccessorImpl.invoke&#40;NativeMethodAccessorImpl.java:62&#41;
	at sun.reflect.DelegatingMethodAccessorImpl.invoke&#40;DelegatingMethodAccessorImpl.java:43&#41;
	at java.lang.reflect.Method.invoke&#40;Method.java:498&#41;
	at com.sun.corba.ee.impl.io.IIOPInputStream.invokeObjectReader&#40;IIOPInputStream.java:1813&#41;
	at com.sun.corba.ee.impl.io.IIOPInputStream.inputObject&#40;IIOPInputStream.java:1195&#41;
	at com.sun.corba.ee.impl.io.IIOPInputStream.simpleReadObject&#40;IIOPInputStream.java:407&#41;
	at com.sun.corba.ee.impl.io.ValueHandlerImpl.readValueInternal&#40;ValueHandlerImpl.java:307&#41;
	at com.sun.corba.ee.impl.io.ValueHandlerImpl.readValue&#40;ValueHandlerImpl.java:273&#41;
	at com.sun.corba.ee.impl.encoding.CDRInputStream_1_0.readRMIIIOPValueType&#40;CDRInputStream_1_0.java:899&#41;
	... 181 more
```

## Steps to reproduce (Only for bug reports) 

Difficult. Same symptoms like  #525.

## Environment ##

- **Payara Version**: 4.1.1.174
- **Edition**: full
- **JDK Version**: 8.0.131
- **Operating System**: Linux
- **Database**: none

--- a/appserver/ejb/ejb-container/src/main/java/com/sun/ejb/EJBUtils.java
+++ b/appserver/ejb/ejb-container/src/main/java/com/sun/ejb/EJBUtils.java
@@ -493,8 +493,12 @@ public class EJBUtils {
         // PAYARA-3087 linkage error occurs when multiple threads load the same generated classes concurrently.
         synchronized (EJBUtils.class) {
 
-            generatedRemoteIntf = loadClassIgnoringExceptions(appClassLoader, generatedRemoteIntfName);
-            generatedRemoteWrapper = loadClassIgnoringExceptions(appClassLoader, wrapperClassName);
+            if (generatedRemoteIntf == null) {
+                generatedRemoteIntf = loadClassIgnoringExceptions(appClassLoader, generatedRemoteIntfName);
+            }
+            if (generatedRemoteWrapper == null) {
+                generatedRemoteWrapper = loadClassIgnoringExceptions(appClassLoader, wrapperClassName);
+            }
 
             if( (generatedRemoteIntf != null) &&
                 (generatedRemoteWrapper != null) ) {

--- a/appserver/ejb/ejb-container/src/main/java/com/sun/ejb/EJBUtils.java
+++ b/appserver/ejb/ejb-container/src/main/java/com/sun/ejb/EJBUtils.java
@@ -482,52 +482,67 @@ public class EJBUtils {
         String wrapperClassName = EJBUtils.
             getGeneratedRemoteWrapperName(businessInterfaceName);
 
-        Class generatedRemoteIntf = null;
-        try {
-            generatedRemoteIntf =
-                appClassLoader.loadClass(generatedRemoteIntfName);
-        } catch(Exception e) {
-        }
-
-        Class generatedRemoteWrapper = null;
-        try {
-            generatedRemoteWrapper =
-                appClassLoader.loadClass(wrapperClassName);
-        } catch(Exception e) {
-        }
+        Class generatedRemoteIntf = loadClassIgnoringExceptions(appClassLoader, generatedRemoteIntfName);
+        Class generatedRemoteWrapper = loadClassIgnoringExceptions(appClassLoader, wrapperClassName);
 
         if( (generatedRemoteIntf != null) &&
             (generatedRemoteWrapper != null) ) {
             return;
         }
 
-        _setClassLoader(appClassLoader);
+        // PAYARA-3087 linkage error occurs when multiple threads load the same generated classes concurrently.
+        synchronized (EJBUtils.class) {
 
+            generatedRemoteIntf = loadClassIgnoringExceptions(appClassLoader, generatedRemoteIntfName);
+            generatedRemoteWrapper = loadClassIgnoringExceptions(appClassLoader, wrapperClassName);
+
+            if( (generatedRemoteIntf != null) &&
+                (generatedRemoteWrapper != null) ) {
+                return;
+            }
+
+            _setClassLoader(appClassLoader);
+
+            try {
+                if( generatedRemoteIntf == null ) {
+
+                    RemoteGenerator gen = new RemoteGenerator(appClassLoader,
+                                                              businessInterfaceName);
+
+                    Class developerClass = appClassLoader.loadClass(businessInterfaceName);
+                    generateAndLoad(gen, generatedRemoteIntfName, appClassLoader, developerClass);
+
+                }
+
+                if( generatedRemoteWrapper == null ) {
+
+                    Remote30WrapperGenerator gen = new Remote30WrapperGenerator
+                        (appClassLoader, businessInterfaceName,
+                         generatedRemoteIntfName);
+
+                    Class developerClass = appClassLoader.loadClass(businessInterfaceName);
+                    generateAndLoad(gen, wrapperClassName, appClassLoader, developerClass);
+                }
+
+            } finally {
+                // Fix for 7075: Make sure no classloader is bound to threadlocal:
+                // avoid possible classloader leak.
+                _setClassLoader(null) ;
+            }
+        }
+    }
+
+    /**
+     * Loads the a class by name using the provided classloader.
+     * @param clsLoader Classloader to use for loading
+     * @param clsName Name of the class to load.
+     * @return loaded class or null in case of an exception.
+     */
+    private static Class loadClassIgnoringExceptions(ClassLoader clsLoader, String clsName) {
         try {
-            if( generatedRemoteIntf == null ) {
-
-                RemoteGenerator gen = new RemoteGenerator(appClassLoader,
-                                                          businessInterfaceName);
-
-                Class developerClass = appClassLoader.loadClass(businessInterfaceName);
-                generateAndLoad(gen, generatedRemoteIntfName, appClassLoader, developerClass);
-
-            }
-
-            if( generatedRemoteWrapper == null ) {
-
-                Remote30WrapperGenerator gen = new Remote30WrapperGenerator
-                    (appClassLoader, businessInterfaceName,
-                     generatedRemoteIntfName);
-
-                Class developerClass = appClassLoader.loadClass(businessInterfaceName);
-                generateAndLoad(gen, wrapperClassName, appClassLoader, developerClass);
-            }
-
-        } finally {
-            // Fix for 7075: Make sure no classloader is bound to threadlocal:
-            // avoid possible classloader leak.
-            _setClassLoader(null) ;
+            return clsLoader.loadClass(clsName);
+        } catch(Exception e) {
+            return null;
         }
     }
 

